src/Compilers/CSharp/Portable/Lowering/LambdaRewriter/LambdaRewriter.Analysis.cs

```diff
@@ -342,6 +341,173 @@ private void MakeAndAssignEnvironments()
                 });
             }
 
+            /// <summary>
+            /// Calculates all closures which directly or indirectly capture a scopes variables.
+            /// </summary>
+            /// <returns></returns>
+            private PooledDictionary<Scope, PooledHashSet<Closure>> CalculateClosuresCapturingScopeVariables()
+            {
+                var closuresCapturingScopeVariables = PooledDictionary<Scope, PooledHashSet<Closure>>.GetInstance();
+
+                // calculate closures which directly capture a scope
+
+                var environmentsToScopes = PooledDictionary<ClosureEnvironment, Scope>.GetInstance();
+
+                VisitScopeTree(ScopeTree, scope =>
+                {
+                    if (scope.DeclaredEnvironments.Count > 0)
+                    {
+                        // Right now we only create one environment per scope
+                        Debug.Assert(scope.DeclaredEnvironments.Count == 1);
+                        var capturingClosures = PooledHashSet<Closure>.GetInstance();
+                        closuresCapturingScopeVariables[scope] = capturingClosures;
+                        environmentsToScopes[scope.DeclaredEnvironments[0]] = scope;
+                    }
+                });
+
+                VisitClosures(ScopeTree, (_, closure) =>
+                {
+                    foreach (var env in closure.CapturedEnvironments)
```

Comment:
Isn't CapturedEnvironments already a transitive list of environments? Why do we need to o the fixup below?